crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1001 @@
+        let components_with_cycles = strongly_connected_components
+            .iter()
+            .filter(|scc| scc.len() > 1)
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let mut message = format!(
+            "schedule contains at least {} cycle(s)",
+            components_with_cycles.len()
+        );
+
+        writeln!(message, " -- cycle(s) found within:").unwrap();
+        for (i, scc) in components_with_cycles.into_iter().enumerate() {
+            let names = scc
+                .iter()
+                .map(|id| self.get_node_name(id))
+                .collect::<Vec<_>>();
+            writeln!(message, " ---- {i}: {names:?}").unwrap();
+        }
+
+        error!("{}", message);
+    }
+
+    fn contains_conflicts(&self, conflicts: &[(NodeId, NodeId, Vec<ComponentId>)]) -> bool {
+        if conflicts.is_empty() {
+            return false;
+        }
+
+        true
+    }
+
+    fn report_conflicts(&self, ambiguities: &[(NodeId, NodeId, Vec<ComponentId>)]) {
+        let mut string = String::from(
+            "Some systems with conflicting access have indeterminate execution order. \
+            Consider adding `before`, `after`, or `ambiguous_with` relationships between these:\n",
+        );
+
+        for (system_a, system_b, conflicts) in ambiguities.iter() {
+            debug_assert!(system_a.is_system());
+            debug_assert!(system_b.is_system());
+            let name_a = self.get_node_name(system_a);
+            let name_b = self.get_node_name(system_b);
+
+            writeln!(string, " -- {name_a} and {name_b}").unwrap();
+            if !conflicts.is_empty() {
+                writeln!(string, "    conflict on: {conflicts:?}").unwrap();
+            } else {
+                // one or both systems must be exclusive
+                let world = std::any::type_name::<World>();
+                writeln!(string, "    conflict on: {world}").unwrap();
+            }
+        }
+
+        warn!("{}", string);
+    }
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum InsertionError {
+    #[error("schedule `{0:?}` already exists")]
+    AlreadyExists(BoxedScheduleLabel),
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum ExtractionError {
+    #[error("unknown schedule: `{0:?}`")]
+    Unknown(BoxedScheduleLabel),
+    #[error("schedule `{0:?}` is not available")]
+    AlreadyExtracted(BoxedScheduleLabel),
+    #[error("schedule `{0:?}` has not been extracted")]
+    NotExtracted(BoxedScheduleLabel),
+}
+
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum BuildError {
+    #[error("`{0:?}` contains itself")]
+    HierarchyLoop(BoxedSystemSet),
+    #[error("system set hierarchy contains cycle(s)")]
+    HierarchyCycle,
+    #[error("system set hierarchy contains conflicting relationships")]
+    HierarchyConflict,
+    #[error("`{0:?}` depends on itself")]
+    DependencyLoop(BoxedSystemSet),
+    #[error("dependencies contain cycle(s)")]
+    DependencyCycle,
```

Comment:
Are we able to surface where the cycle was detected in any way? This may be nasty to find.

Reply:
Possibly.

It'll log the strongly connected components that have cycles right now, but there could >1 cycle in each one. I think there's an algorithm that can enumerate the simple cycles in each SCC.

Reply:
Yep, totally happy to leave this to a follow-up. Agreed on `bevy_graph`!